### PR TITLE
reverting camera dist

### PR DESCRIPTION
**Pull request by GitWillG**
When unaiming, you now revert back to the original camera distance.

### CODE REVIEW

File: camera-manager.js [hunk 1]
```diff
@@ -5,6 +5,8 @@ import metaversefile from 'metaversefile';
 import physicsManager from './physics-manager.js';
 import {shakeAnimationSpeed} from './constants.js';
 import Simplex from './simplex-noise.js';
+import { lerp } from 'three/src/math/MathUtils';
+import { localPlayer } from './players.js';
```
Comment (Loryhoof) [inline camera-manager.js:9]: Hmm, that works but I'm not sure if it's intended to be used like that.
You can get the localPlayer like this:
`const localPlayer = metaversefileApi.useLocalPlayer();`

File: camera-manager.js [hunk 2]
```diff
@@ -206,6 +217,25 @@ class CameraManager extends EventTarget {
     }
     return result;
   }
+  
+  //Lerp towards offset for smoother ADS, works both aiming in and out
+  ads(){
```
Comment (Loryhoof) [inline camera-manager.js:222]: Should probably change "ads" to something more intuitive

File: camera-manager.js [hunk 3]
```diff
@@ -206,6 +217,25 @@ class CameraManager extends EventTarget {
...
+    if (localPlayer.hasAction('aim') && -cameraOffset.z>1){
```
Comment (Loryhoof) [inline camera-manager.js:223]: `-cameraOffset.z>1` should probably not be hardcoded, could add it as a variable above so it's easier to edit.

File: .vscode/launch.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (avaer) [inline .vscode/launch.json:1]: Probably shouldn't commit this.

File: camera-manager.js (hunk 1 above)
Comment (avaer) [reply camera-manager.js:9]: It's fine to import `localPlayer` like this if you are not an app.

File: game.js [hunk 5]
```diff
@@ -1154,6 +1154,10 @@ class GameManager extends EventTarget {
   menuAim() {
     const localPlayer = metaversefileApi.useLocalPlayer();
     if (!localPlayer.hasAction('aim')) {
+      //current ads value is just a hardcoded offset value that seemed appropriate
+      cameraManager.saveaim();
```
Comment (avaer) [inline game.js:1158]: Is this missing from this PR?

File: camera-manager.js (hunk 1 above)
Comment (GitWillG) [reply camera-manager.js]: I just entirely removed the dependency on localplayer from within the aim functionality. I only need to pass in the players position once at the start of takeaim anyways

File: camera-manager.js (hunk 3 above)
Comment (GitWillG) [reply camera-manager.js]: This was an oversight while I was testing different distances, but it now uses the inverted value from "maxAim" an already existing variable above

File: game.js (hunk 5 above)
Comment (GitWillG) [reply game.js:1157]: I actually removed it entirely in a more recent PR

File: camera-manager.js [hunk 6]
```diff
@@ -355,13 +402,23 @@ class CameraManager extends EventTarget {
               localVector.copy(avatarCameraOffset)
                 .applyQuaternion(camera.quaternion)
             );
+
+            //only aims down sight when not in first person mode
+            if (localPlayer.hasAction('aim')){
+            this.takeAim(localPlayer.position);
```
Comment (avaer) [inline camera-manager.js:408]: Spacing issues.

File: camera-manager.js [hunk 7]
```diff
@@ -206,6 +209,40 @@ class CameraManager extends EventTarget {
     }
     return result;
   }
+  
+  //Lerp towards offset for smoother ADS, works both aiming in and out
+  takeAim(playerPos){
+    if (-cameraOffset.z> (-maxAim.z)){
```
Comment (Loryhoof) [inline camera-manager.js:215]: No need for parentheses `( )`

File: camera-manager.js [hunk 8]
```diff
@@ -206,6 +209,40 @@ class CameraManager extends EventTarget {
...
+    //we only want to revert to our original z value if there is a descrepancy between our expected zdistance and our actual distance
+    if (zdist !== 0){
+      cameraOffsetZ = lerpNum(cameraOffsetZ, lastCamoffsetz, 0.1);
+      cameraOffsetTargetZ = cameraOffsetZ;
+    }
+    else{
+      //do nothing
+    }
+    cameraOffset.x = lerp(cameraOffset.x, rayVectorZero.x, 0.1);
+    cameraOffset.updateMatrixWorld;
```
Comment (Loryhoof) [inline camera-manager.js:243]: `updateMatrixWorld` doesn't do anything here

File: camera-manager.js [hunk 9]
```diff
@@ -15,8 +16,12 @@ const localEuler = new THREE.Euler();
 
 const cameraOffset = new THREE.Vector3();
 let cameraOffsetTargetZ = cameraOffset.z;
+//
+const maxAim = new THREE.Vector3(-0.3,0,-2);
 
+//
 let cameraOffsetZ = cameraOffset.z;
+let lastCamoffsetz = cameraOffset.z;
```
Comment (Loryhoof) [inline camera-manager.js:24]: make it capital Z

File: camera-manager.js [hunk 10]
```diff
@@ -206,6 +209,40 @@ class CameraManager extends EventTarget {
...
+    if (-cameraOffset.z> -maxAim.z){
+      localVector.copy(playerPos).sub(localVector.copy(maxAim).applyQuaternion(camera.quaternion));
+      cameraOffsetZ = lerpNum(cameraOffsetZ, -localVector.z, 0.1);
+      cameraOffsetTargetZ = cameraOffsetZ;
+    }
+    else{
+      //do nothing
+    }
+    cameraOffset.x = lerp(cameraOffset.x, maxAim.x, 0.1);
+    cameraOffset.updateMatrixWorld;
```
Comment (Loryhoof) [inline camera-manager.js:224]: `updateMatrixWorld` doesn't do anything here either

File: game.js [hunk 11]
```diff
@@ -1154,6 +1154,7 @@ class GameManager extends EventTarget {
   menuAim() {
     const localPlayer = metaversefileApi.useLocalPlayer();
     if (!localPlayer.hasAction('aim')) {
+      cameraManager.saveaim();
```
Comment (Loryhoof) [inline game.js:1157]: small detail but try to keep all variable/function names congruent. In this case camelcase

File: camera-manager.js (hunk 7 above)
Comment (Loryhoof) [reply camera-manager.js]: and the spacing in `z>`

**Comment by avaer** [comment]
This doesn't interact very well if you scroll while holding aim. It will just jitter.

**Comment by avaer** [comment]
Due to this PR being out of date from master it cannot be tested (will crash on sprite avatar settings for example).

**Comment by avaer** [comment]
image

**Comment by msm0823** [comment]
As far as the functionality is concerned, it is working fine. The camera does reverts back to its original position while unaiming. Also, the limited transitions seems fine when we tried to scroll while aiming.

A little issue that seems to be here is the aiming point. When we aim at smaller objects, such as a potion we have to aim a little right side to target it actually.

**Pull request closed without merging**